### PR TITLE
Theme audio for all GEL buttons (ASR and WW themes)

### PR DESCRIPTION
**Pull request by AdamXA**
General polish - theme audio for all GEL buttons.

Had to make minor changes to PauseManager, Pause and How to Play screens to get this to work. Ideally, these need to be rethought about in future so that this is much tidier and simple to do.

### CODE REVIEW

File: src/components/how-to-play/how-to-play.ts [hunk 1]
```diff
@@ -70,7 +81,11 @@ export function create(game: Phaser.Game, screen: Screen) {
     }
 
     function destroy() {
+        if (screen.pauseManager) {
+            screen.pauseManager.destroyButtons("howToPlay");
+        }
         layout.destroy();
+        background.kill();
```
Comment (FostUK) [inline src/components/how-to-play/how-to-play.ts]: Probably only need one of these.

File: src/components/how-to-play/how-to-play.ts [hunk 2]
```diff
@@ -32,18 +32,25 @@ export function create(game: Phaser.Game, screen: Screen) {
...
-        pauseManager.pauseifyButton(layout.buttons.next);
-        pauseManager.pauseifyButton(layout.buttons.previous);
-    }
+    const layout = context.gel.createLayout(["previous", "next", "back"]);
 
+    if (game.paused && screen.pauseManager) {
+        screen.pauseManager.setBackground(background);
+        for (const key in layout.buttons) {
+            if (key) {
+                screen.pauseManager.pauseifyButton(layout.buttons[key], sfx, key, "howToPlay");
```
Comment (SeanHolden) [inline src/components/how-to-play/how-to-play.ts]: I don't feel like this loop is necessary. It's less obvious what is happening and we now have no control over which buttons are passed in.

File: src/components/pause/pause.ts [hunk 3]
```diff
@@ -1,35 +1,30 @@
...
-    pauseManager.pauseifyButton(layout.buttons.howToPlay, howToPlaySfx);
-    pauseManager.pauseifyButton(layout.buttons.restart);
-    pauseManager.pauseifyButton(layout.buttons.home);
-    pauseManager.pauseifyButton(layout.buttons.play);
-
+    const sfx: Phaser.AudioSprite = game.add.audioSprite(context.gel.keyLookup.sfx);
+
+    screen.pauseManager = PauseManager.create(game, screen.context);
+    screen.pauseManager.setBackground(background);
+    for (const key in layout.buttons) {
```
Comment (FostUK) [inline src/components/pause/pause.ts]: and maybe _.forOwn here maybe too.

File: src/components/how-to-play/how-to-play.ts [hunk 4]
```diff
@@ -32,18 +32,25 @@ export function create(game: Phaser.Game, screen: Screen) {
...
+            }
+        }
+    } else {
+        for (const key in layout.buttons) {
+            if (key) {
+                layout.buttons[key].onInputOver.add(onOver, undefined, undefined, key);
```
Comment (SeanHolden) [inline src/components/how-to-play/how-to-play.ts]: Why do we need to do this?

File: src/components/results/results.ts [hunk 5]
```diff
@@ -22,9 +24,16 @@ class Results extends Screen {
         this.layout.setAction("restart", this.restartGame.bind(this));
         this.layout.setAction("continue", this.continue.bind(this));
         this.layout.setAction("pause", () => {
-            this.pause = Pause.create(this.game, this);
+            Pause.create(this.game, this);
         });
         this.layout.setAction("home", this.goHome.bind(this));
+
+        for (const key in this.layout.buttons) {
```
Comment (FostUK) [inline src/components/results/results.ts]: _.forOwn here too.

File: src/components/pause/pause.ts [hunk 6]
```diff
@@ -1,35 +1,30 @@
...
+        if (key) {
+            screen.pauseManager.pauseifyButton(layout.buttons[key], sfx, key, "pause");
```
Comment (SeanHolden) [inline src/components/pause/pause.ts]: It's a shame we have to pass in the "pause" string on every `pausifyButton` now. Is there no other way?

File: src/core/pause-manager.ts [hunk 7]
```diff
@@ -1,23 +1,36 @@
...
-    pauseifyButton: (button: Phaser.Button, hoverSfx?: Phaser.Sound) => PauseifiedButton;
+    update(): void;
+    getButtons(popupScreenContext: string): PauseifiedButton[];
+    destroyButtons(popupScreenContext: string): void;
+    setBackground(background: Phaser.Image): void;
+    pauseifyButton(
+        button: Phaser.Button,
+        hoverAudioSprite: Phaser.AudioSprite,
+        marker: string,
+        popupScreenContext: string,
```
Comment (SeanHolden) [inline src/core/pause-manager.ts]: Passing this argument in here seems like a code smell. Every button that gets "pauseified" now needs it. Is there definitely no way we can just pass it in once to an instance of pauseManager? Which then pauseifyButton is called from.

**Comment by SeanHolden** [comment]
There's a lot of code/assets in here not related to the Title of this PR and there is no description. It's not a huge deal as this is a prototype but for future PRs, we should really be keeping our PRs simple, descriptive and in scope.

File: src/components/title/title.ts [hunk 8]
```diff
@@ -5,27 +5,37 @@ import { Screen } from "core/screen";
...
         this.layout = gel.createLayout(["exit", "howToPlay", "play", "soundOn", "settings"]);
-
         this.layout.addToGroup("middleCenterV", this.make.sprite(0, 0, theme.keyLookup.logo), 0);
-
         this.layout.setAction("play", this.startGame.bind(this));
         this.layout.setAction("howToPlay", () => {
             HowToPlay.create(this.game, this);
         });
 
+        for (const key in this.layout.buttons) {
```
Comment (FostUK) [inline src/components/title/title.ts]: _.forOwn here

File: src/components/how-to-play/how-to-play.ts [hunk 9]
```diff
@@ -55,6 +58,10 @@ export function create(game: Phaser.Game, screen: Screen) {
         update();
     }
 
+    function onOver(button: Phaser.Button, pointer: Phaser.Pointer, marker: string) {
```
Comment (SeanHolden) [inline src/components/how-to-play/how-to-play.ts]: Can we change the name of this function to be more descriptive? For example `playSound` perhaps?

File: src/game/game.ts [hunk 10]
```diff
@@ -156,9 +158,15 @@ export class Racer extends Screen {
 
         this.layout = this.context.gel.createLayout(["home", "pause"]);
         this.layout.setAction("pause", () => {
+            this.game.sound.pauseAll();
             Pause.create(this.game, this);
         });
         this.layout.setAction("home", this.goHome.bind(this));
+        for (const key in this.layout.buttons) {
```
Comment (FostUK) [inline src/game/game.ts]: _.forOwn might be good here

File: src/core/pause-manager.ts (hunk 7 above)
Comment (AdamXA) [reply src/core/pause-manager.ts]: Ideally I think we need some kind of Layer manager. Didn't want to refactor the whole pauseManager, my changes have been as minimal as possible to get the audio working on all the GEL buttons.

**Comment by hannahbunce** [comment]
There's no audio on the WW buttons.

**Comment by hannahbunce** [comment]
There's now audio on the Worst Witch GEL buttons. Good to merge!

**Pull request merged**